plugins/inputs/prometheus_scraper/metrics_receiver.go

```diff
@@ -106,7 +64,71 @@ func (ma *metricAppender) Rollback() error {
 	return nil
 }
 
-func (ma *metricAppender) AppendExemplar(ref storage.SeriesRef, l labels.Labels, e exemplar.Exemplar) (storage.SeriesRef, error) {
-	ma.Append(ref, l, e.Ts, e.Value)
+func (ma *metricAppender) AppendExemplar(ref storage.SeriesRef, ls labels.Labels, e exemplar.Exemplar) (storage.SeriesRef, error) {
+	ma.Append(ref, ls, e.Ts, e.Value)
 	return 0, nil
 }
+
+func (ma *metricAppender) AppendMetricToBatch(ls labels.Labels, metricCreateTime int64, metricValue float64) (err error) {
+	// Each new scrape will create a context hold metadataStore. Therefore, the same context will be used
+	// by all metrics in the same batch. So we only need to fetch the metadataStore once
+	if ma.isNewBatch {
+		metadataCache, err := getMetadataCache(ma.ctx)
+		if err != nil {
+			return err
+		}
+		ma.isNewBatch = false
+		ma.mc = metadataCache
+	}
```

Comment:
Why not just update the `Appender` function to get the metadata cache when creating the appender?